File: tests/fixtures/cpp_repo/src/handlers/users_handler.cpp
#include "api/handlers/users_handler.hpp"

#include "api/dto/user_dto.hpp"
#include "api/error/api_error.hpp"
#include "api/services/user_service.hpp"

#include <httplib.h>
#include <nlohmann/json.hpp>

#include <cstdlib>
#include <utility>

namespace api {

namespace {

void write_json(httplib::Response& res, int status, const nlohmann::json& body) {
    res.status = status;
    res.set_content(body.dump(), "application/json");
}

void write_error(httplib::Response& res, const ApiError& err) {
    nlohmann::json envelope = {
        {"error", {{"code", err.code()}, {"message", err.message()}}}
    };
    write_json(res, err.status(), envelope);
}

std::int64_t parse_id(const httplib::Request& req) {
    try {
        return std::stoll(req.matches[1]);
    } catch (const std::exception&) {
        throw ApiError::bad_request("invalid id segment");
    }
}

int parse_query_int(const httplib::Request& req, const char* key, int fallback) {
    auto it = req.params.find(key);
    if (it == req.params.end()) {
        return fallback;
    }
    try {
        return std::stoi(it->second);
    } catch (const std::exception&) {
        return fallback;
    }
}

}  // namespace

UsersHandler::UsersHandler(std::shared_ptr<UserService> service)
    : service_(std::move(service)) {}

void UsersHandler::create_user(const httplib::Request& req, httplib::Response& res) {
    try {
        auto body = nlohmann::json::parse(req.body);
        auto request = CreateUserRequest::from_json(body);
        UserResponse created = service_->create_user(request);
        write_json(res, 201, created.to_json());
    } catch (const ApiError& err) {
        write_error(res, err);
    } catch (const nlohmann::json::exception&) {
        write_error(res, ApiError::bad_request("malformed json"));
    }
}

void UsersHandler::get_user(const httplib::Request& req, httplib::Response& res) {
    try {
        UserResponse user = service_->get_user_by_id(parse_id(req));
        write_json(res, 200, user.to_json());
    } catch (const ApiError& err) {
        write_error(res, err);
    }
}

void UsersHandler::list_users(const httplib::Request& req, httplib::Response& res) {
    try {
        int page = parse_query_int(req, "page", 0);
        int page_size = parse_query_int(req, "page_size", 20);
        auto rows = service_->list_users(page, page_size);
        nlohmann::json array = nlohmann::json::array();
        for (const auto& row : rows) {
            array.push_back(row.to_json());
        }
        write_json(res, 200, array);
    } catch (const ApiError& err) {
        write_error(res, err);
    }
}

void UsersHandler::update_user(const httplib::Request& req, httplib::Response& res) {
    try {
        auto body = nlohmann::json::parse(req.body);
        auto patch = UpdateUserRequest::from_json(body);
        UserResponse updated = service_->update_user(parse_id(req), patch);
        write_json(res, 200, updated.to_json());
    } catch (const ApiError& err) {
        write_error(res, err);
    } catch (const nlohmann::json::exception&) {
        write_error(res, ApiError::bad_request("malformed json"));
    }
}

void UsersHandler::delete_user(const httplib::Request& req, httplib::Response& res) {
    try {
        service_->delete_user(parse_id(req));
        res.status = 204;
    } catch (const ApiError& err) {
        write_error(res, err);
    }
}

}  // namespace api

File: tests/fixtures/go_repo/internal/handlers/users_handler.go
// Package handlers — /users CRUD endpoints.
package handlers

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"

	"github.com/example/goapi/internal/repository"
	"github.com/example/goapi/internal/services"
	"github.com/example/goapi/internal/types"
)

// UsersHandler hosts /users endpoints.
type UsersHandler struct {
	users *services.UserService
}

// NewUsersHandler constructs a UsersHandler.
func NewUsersHandler(users *services.UserService) *UsersHandler {
	return &UsersHandler{users: users}
}

// CreateUser handles POST /users. Returns 201 with the created user on
// success, 409 when the email is already taken.
func (h *UsersHandler) CreateUser(w http.ResponseWriter, r *http.Request) {
	var req types.CreateUserRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	user, err := h.users.CreateUser(r.Context(), &req)
	if err != nil {
		if errors.Is(err, services.ErrEmailAlreadyTaken) {
			writeError(w, http.StatusConflict, "email already taken")
			return
		}
		writeError(w, http.StatusInternalServerError, "could not create user")
		return
	}
	writeJSON(w, http.StatusCreated, toUserResponse(user))
}

// GetUser handles GET /users/{id}.
func (h *UsersHandler) GetUser(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	user, err := h.users.GetUserByID(r.Context(), id)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			writeError(w, http.StatusNotFound, "user not found")
			return
		}
		writeError(w, http.StatusInternalServerError, "could not fetch user")
		return
	}
	writeJSON(w, http.StatusOK, toUserResponse(user))
}

// ListUsers handles GET /users with page/page_size query params.
func (h *UsersHandler) ListUsers(w http.ResponseWriter, r *http.Request) {
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	pageSize, _ := strconv.Atoi(r.URL.Query().Get("page_size"))
	rows, total, err := h.users.ListUsers(r.Context(), page, pageSize)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "could not list users")
		return
	}
	resp := types.UserListResponse{
		Items:      make([]types.UserResponse, 0, len(rows)),
		TotalCount: total,
		Page:       page,
		PageSize:   pageSize,
	}
	for i := range rows {
		resp.Items = append(resp.Items, toUserResponse(&rows[i]))
	}
	writeJSON(w, http.StatusOK, resp)
}

// UpdateUser handles PATCH /users/{id}.
func (h *UsersHandler) UpdateUser(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var patch types.UpdateUserRequest
	if err := json.NewDecoder(r.Body).Decode(&patch); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	user, err := h.users.UpdateUser(r.Context(), id, &patch)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			writeError(w, http.StatusNotFound, "user not found")
			return
		}
		writeError(w, http.StatusInternalServerError, "could not update user")
		return
	}
	writeJSON(w, http.StatusOK, toUserResponse(user))
}

// DeleteUser handles DELETE /users/{id}.
func (h *UsersHandler) DeleteUser(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	if err := h.users.DeleteUser(r.Context(), id); err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			writeError(w, http.StatusNotFound, "user not found")
			return
		}
		writeError(w, http.StatusInternalServerError, "could not delete user")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
